### PR TITLE
FHEUIntX type boilerplate code

### LINKED ISSUE #20: Generate boilerplate code for Solidity FHEUInt types

### PR DESCRIPTION
**Pull request by tremblaythibaultl**
Introduce a simple python script to generate the FHElib code.

This boilerplate code will let Solidity developer get compile-time type checking with FHEUInt types as well as runtime checking.

Closes #20.

### CODE REVIEW

File: codegen/FHEOps.sol [hunk 1]
```diff
@@ -0,0 +1,906 @@
...
+pragma solidity >=0.8.13 <0.9.0;
+
+type euint8 is uint256;
+type euint16 is uint256;
+type euint32 is uint256;
+type euint64 is uint256;
+type euint128 is uint256;
+type euint256 is uint256;
+
+library Common {
```
Comment (dartdart26) [inline codegen/FHEOps.sol]: We have a `Common` library in the Common.sol file. Do these become part of the same library?

File: codegen/FHEOps.sol [hunk 2]
```diff
@@ -0,0 +1,906 @@
...
+    // Values used to communicate types at runtime to the cast() precompile.
```
Comment (dartdart26) [inline codegen/FHEOps.sol]: Got me thinking about whether we should also pass the compile-time types to the runtime in operations too. I think we discussed it, but I forgot what we decided at the end.

File: codegen/FHEOps.sol [hunk 3]
```diff
@@ -0,0 +1,906 @@
...
+
+    function to_euint8(euint16 v) internal view returns (euint8) {
+        return euint8.wrap(Impl.cast(euint16.unwrap(v), Common.euint16_t));
+    }
+
+    function to_euint8(euint32 v) internal view returns (euint8) {
+        return euint8.wrap(Impl.cast(euint32.unwrap(v), Common.euint32_t));
+    }
+
+    function to_euint8(euint64 v) internal view returns (euint8) {
```
Comment (dartdart26) [inline codegen/FHEOps.sol]: Should we call these `toEuint`?

**Comment by dartdart26** [comment]
A thing that comes to mind is that we need to be explicit and document the behaviour on integer overflow. AFAIK, at the moment, we cannot detect it for euint types and make them behave as uint types. Even if we can detect it in the future, we'd have to decrypt a boolean that denotes whether overflow happened. Maybe we could have two versions of these operations - one that checks and one that doesn't.

File: codegen/FHEOps.sol (hunk 1 above)
Comment (tremblaythibaultl) [reply codegen/FHEOps.sol]: I think we can make a split (one file per library) as we used to.

**Comment by tremblaythibaultl** [comment]
> AFAIK, at the moment, we cannot detect it for euint types and make them behave as uint types. 

To be clear, euint types behave like "regular" `uint`s with wrap-around-on-overflow logic, not solidity uints with revert-on-overflow logic.

**Comment by tremblaythibaultl** [comment]
Closes #20 and is made to be used with https://github.com/zama-ai/go-ethereum/issues/63.

File: types/TypedEncryptedERC20.sol [hunk 4]
```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+
+pragma solidity >=0.8.13 <0.9.0;
+
+import "./types/lib/Ciphertext.sol";
+import "./types/lib/Common.sol";
+import "./types/lib/FHEOps.sol";
+
+contract TypedEncryptedERC20 {
```
Comment (dartdart26) [inline types/TypedEncryptedERC20.sol]: Should we make all code in the examples folder work with the new library as part of this PR? Say just change from FHEUInt to euint16 or whatever type?

Comment (tremblaythibaultl) [reply types/TypedEncryptedERC20.sol]: Sure, I'll add an issue for this.

**Comment by tremblaythibaultl** [comment]
Commit [5788b2d](https://github.com/zama-ai/zbc-solidity/pull/23/commits/5788b2dd252391345177b662c509ffbad577d184) closes #27.

File: codegen/FHEOps.sol (hunk 3 above)
Comment (tremblaythibaultl) [reply codegen/FHEOps.sol]: Done in [daa4e0b](https://github.com/zama-ai/zbc-solidity/pull/23/commits/daa4e0b67bb405af93713400de789ab5f758c42a).

File: types/TypedEncryptedERC20.sol (hunk 4 above)
Comment (tremblaythibaultl) [reply types/TypedEncryptedERC20.sol]: Done in [5788b2d](https://github.com/zama-ai/zbc-solidity/pull/23/commits/5788b2dd252391345177b662c509ffbad577d184).

**Pull request merged**